A51-custom-backend-metrics.md

```diff
@@ -0,0 +1,462 @@
+  // Create a bindable service and allow configuring minimum report interval.
+  // If not configured or badly configured non-positive,
+  // the default is 30s. Bind the returned service to the server.
+  public BindableService createService(ScheduledExecutorService timeService, long minReportInterval,
+                                     TimeUnit unit);
+
+  // In the future, it might be useful to let users specify minimum report interval and default 
+  // report interval separately. If the default report interval is not configured or badly configured 
+  // non-positive, the default is 1 min. 
+  // public BindableService createService(long minInterval, TimeUnit minUnit,
+  // long defaultInterval, TimeUnit defaultUnit)
+
+  // Creates an OOB bindable service using default minimum report interval = 30s.
+  public BindableService createService(ScheduledExecutorService timeService);
+
+  // Update the metrics value corresponding to the specified key.
+  public void setUtilizationMetric(String key, double value);
+
+  // Replace the whole metrics data using the specified map.
+  public void setAllUtilizationMetrics(Map<String, Double> metrics);
+
+  // Remove the metrics data entry corresponding to the specified key.
+  public void deleteUtilizationMetric(String key);
+
+  // Update the CPU utilization metrics data.
+  public void setCPUUtilizationMetric(double value);
+
+  // Clear the CPU utilization metrics data.
+  public void deleteCPUUtilizationMetric();
+
+  // Update the memory utilization metrics data.
+  public void setMemoryUtilizationMetric(double value);
+
+  // Clear the memory utilization metrics data.
+  public void deleteMemoryUtilizationMetric();
+}
+```
+
+At the client side, Java will provide a utility function to wrap the `LoadBalancer.Helper`,
+usable by both WRR policy and xDS LB policies. The `LoadBalancer.Helper` manages the OOB stream
+lifecycle and accepts subscriptions from any routing hierarchy.
+The OOB stream will also gracefully disable itself if the server integration is not complete, 
+so users do not have to worry about client and server deployment coordination.
+
+The following pseudocode shows how a custom load balancing policy instantiates the helper wrapper, 
+installs the out-of-band metrics reporting mechanism, and then consumes the metrics reports, 
+which only requires users to register an `OrcaOobReportListener` corresponding to the subchannel.
+
+```java
+class CustomLoadBalancer extends LoadBalancer { 
+  private final Helper helper;  // the original Helper
+  void handleNewServer(ResolvedAddresses address) {
+    // Create OrcaHelperWrapper and a separate listener for each Subchannel
+    OrcaListener listener = new OrcaOobReportListener();
+    OrcaHelperWrapper orcaWrapper = OrcaUtil.newOrcaHelperWrapper(helper);
+    Subchannel subchannel = orcaWrapper.asHelper().create(
+        CreateSubchannelArgs.newBuilder().setAddresses(address).build());
+    orcaWrapper.setListener(subchannel, listener,
```

Comment:
I really think this should be static, as I can't find any value in using the `orcaWrapper` and the static method is easier in the face of composition (if another helper needs to wrap the orca oob).